src/lib/core/interval.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include <stdint.h>
+#include <string.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+enum {
+	/**
+	 * The length of the buffer sufficient to contain the contents of any
+	 * string representation of value of struct interval.
+	 */
+	INTERVAL_STR_MAX_LEN = 256,
+};
+
+/**
+ * Type of adjust of date interval.
+ *
+ * In some cases, when the original date is increased or decreased by years or
+ * months, the result date may not have the same number as the original date.
+ * For example, when 1 year is added to February 29 or 1 month is subtracted
+ * from December 31. This option tells what the result will be in such cases.
+ */
+enum adjust {
+	/** See the description of DT_EXCESS in c-dt. */
+	INTERVAL_ADJUST_EXCESS,
+	/** See the description of DT_LIMIT in c-dt. */
+	INTERVAL_ADJUST_LIMIT,
+	/** See the description of DT_SNAP in c-dt. */
+	INTERVAL_ADJUST_SNAP,
+};
+
+/** A structure that describes a date and time interval. */
+struct interval {
+	/** Number of years. */
+	int32_t year;
+	/** Number of months. */
+	int32_t month;
+	/** Number of weeks. */
+	int32_t week;
+	/** Number of days. */
+	int32_t day;
+	/** Number of hours. */
+	int32_t hour;
+	/** Number of minutes. */
+	int32_t min;
+	/** Number of seconds. */
+	int32_t sec;
```

Comment:
Here is the explanation why it's `double` in #6923 :
 - let assume we are subtracting 2 datetime with fields `epoch` (which is `double`) and `nsec` (which is `int`);
 - what type of result should be for '+' or '-' between 2 `epoch`s?

`int` is not sufficient enough to handle those values we could save there, only `int64_t` (which is boxed and produces performance impact due to GC) or `double` (which is unboxed, and native for Lua).

That's why we have to use same type as in `epoch`. 

OTOH, we could try to decompress full seconds difference into `year`, '`month`, 'day`, etc. But in this case __every__ arithmetic operation would require full cycle of normalization (including leap seconds, leap years, timezone differences, and what not) which is absolutely unnecessary here, and might be entirely avoided if simply save difference with `epoch` type...

Reply:
We shouldn't subtract one epoch from another directly.